processing/src/main/java/io/druid/query/groupby/GroupByQuery.java

```diff
@@ -143,54 +137,48 @@ public GroupByQuery(
     // We're not counting __time, even though that name is problematic. See: https://github.com/druid-io/druid/pull/3684
     verifyOutputNames(this.dimensions, this.aggregatorSpecs, this.postAggregatorSpecs);
 
+    limitFn = makeLimitFn(this.limitSpec);
+  }
+
+  private Function<Sequence<Row>, Sequence<Row>> makeLimitFn(LimitSpec limitSpec)
+  {
     Function<Sequence<Row>, Sequence<Row>> postProcFn =
-        this.limitSpec.build(this.dimensions, this.aggregatorSpecs, this.postAggregatorSpecs);
+        limitSpec.build(dimensions, aggregatorSpecs, postAggregatorSpecs);
 
     if (havingSpec != null) {
       postProcFn = Functions.compose(
           postProcFn,
-          new Function<Sequence<Row>, Sequence<Row>>()
-          {
-            @Override
-            public Sequence<Row> apply(Sequence<Row> input)
-            {
-              GroupByQuery.this.havingSpec.setRowSignature(GroupByQueryHelper.rowSignatureFor(GroupByQuery.this));
-              return Sequences.filter(
-                  input,
-                  new Predicate<Row>()
-                  {
-                    @Override
-                    public boolean apply(Row input)
-                    {
-                      return GroupByQuery.this.havingSpec.eval(input);
-                    }
-                  }
-              );
-            }
+          (Sequence<Row> input) -> {
+            havingSpec.setRowSignature(GroupByQueryHelper.rowSignatureFor(GroupByQuery.this));
+            return Sequences.filter(input, havingSpec::eval);
           }
       );
     }
+    return postProcFn;
+  }
 
-    limitFn = postProcFn;
```

Comment:
While you're at it, this should probably be called `this.postProcFn` since it's doing both LIMIT and HAVING.